### PR TITLE
Disable fleets bootstrap

### PR DESCRIPTION
**Pull request by manno**
## Problem

Both Fleet and Rancher try to manage the local cluster and overwrite each others changes.
 
* labels are overwritten, e.g. when controller pods restart, bundles might get uninstalled if they rely on labels
* local clusters uses a different API endpoint than downstream clusters, downstream clusters use `internal-server-url`
* the local clusters kubeconfig is not updated when Rancher changes
* we cannot move the local cluster to another cluster, as the other controller would create a second one
* ownership and clean up of the local cluster differs between the two controllers
* ...

## Solution

Only Rancher will manage the local cluster. "Bootstrapping" is disabled in Fleet (the "local cluster" block from https://fleet.rancher.io/ref-registration).

The https://github.com/rancher/rancher/blob/release/v2.7/pkg/controllers/provisioningv2/fleetcluster/fleetcluster.go already bootstraps the local cluster.
If both manage the same cluster resource. This is a problem, because the
controllers will overwrite each others changes as the owner information differs.

By using Rancher's bootstrapping, we no longer have the bootstrap
resources from Fleet, like the local cluster group and the local cluster kubeconfig.

Creation of the local cluster group is added to the fleetcluster controller.

Rancher already creates a 'local-kubeconfig' secret for Fleet, which was not used previously.
This secret contains the "CATTLE_SERVER_URL" as API server, but Fleet's
secret contained the internal address of the API server.
By switching to the "CATTLE_SERVER_URL", the local cluster can no longer
be setup without a working URL.

Note: There is an additional token field in 'local-kubeconfig', which is
not used by Fleet (see `importCluster`).

The local cluster was switched in [5b9fbac49](https://github.com/rancher/rancher/commit/5b9fbac497d17dfe11e4ac6857032f628218df52) to Fleet's secret. That issue should not occur anymore, since only one controller manages the resource.

### Manual Testing

This was tested manually by running Rancher in debug mode.

### Automated Testing

Added a unit test and itegration test.

## QA Testing Considerations

 
Does that work with all kinds of API server URLs. Is the assumption correct, that if the API server cannot be reached, we have larger problems than fleet not registering its local cluster?

### Regressions Considerations

Instead of using the public IP, if it is necessary to handle the 
"loopback connect is handled different from any other cluster" we might need to bring Fleet's `buildKubeconfig` code into the Rancher controller.

### CODE REVIEW

**Comment by manno** [comment]
Not sure why CI fails.
Running the tests locally, the fleet-agent cannot connect to the cluster's k8s API, because of an invalid CA.

And indeed, the CA from fleet's config is different from the one server on port 8443. The key and serial number differs.

**Review by jakefhyde — APPROVED** [review]

File: pkg/controllers/provisioningv2/fleetcluster/fleetcluster.go
Comment (Oats87) [inline pkg/controllers/provisioningv2/fleetcluster/fleetcluster.go:145]: Is there the possibility that a user that is upgrading has not already created a `ClusterGroup` called `default` in the `fleet-local` namespace?

Do we not anticipate the potential for a conflict in name here?

`cluster.Namespace` will be `fleet-local` in this case, so we're looking at a cluster group called for `fleet-local/default`.

File: pkg/controllers/provisioningv2/fleetcluster/fleetcluster.go
Comment (Oats87) [inline pkg/controllers/provisioningv2/fleetcluster/fleetcluster.go:139]: We probably don't need this anymore, the `status.ClientSecretName` for the `fleet-local/local` cluster should already be `local-kubeconfig`

Comment (Oats87) [reply pkg/controllers/provisioningv2/fleetcluster/fleetcluster.go:139]: This is a nit though, I don't see harm in duplicating this

File: pkg/controllers/provisioningv2/fleetcluster/fleetcluster.go
Comment (Oats87) [inline pkg/controllers/provisioningv2/fleetcluster/fleetcluster.go:49]: just out of curiosity, why are you defining this function dynamically? I may be missing the context behind this.

Comment (Oats87) [reply pkg/controllers/provisioningv2/fleetcluster/fleetcluster.go:49]: nevermind I see the test

File: tests/integration/pkg/tests/fleet/fleetcluster_test.go
Comment (Oats87) [inline tests/integration/pkg/tests/fleet/fleetcluster_test.go:25]: Would you mind moving this to the new location i.e. `tests/v2prov/pkg/tests` and renaming the test so that it runs in CI?

https://github.com/rancher/rancher/tree/release/v2.7/tests/v2prov#general for more context on the new location

Comment (Oats87) [reply tests/integration/pkg/tests/fleet/fleetcluster_test.go:25]: If this is merged as it stands it won't be run by anything.

**Review by Oats87 — APPROVED** [review]
For the sake of knowing dev complete is today, I am giving this an approval; however, I did have two comments on it (namely around cluster group name) and around the new integration test.

If you want to address those two before merging, then I am happy to re-approve as necessary.

File: pkg/controllers/provisioningv2/fleetcluster/fleetcluster.go
Comment (manno) [reply pkg/controllers/provisioningv2/fleetcluster/fleetcluster.go:49]: Yes, that's the reason I was hesitant to write a unit test in the first place.

File: tests/integration/pkg/tests/fleet/fleetcluster_test.go
Comment (manno) [reply tests/integration/pkg/tests/fleet/fleetcluster_test.go:25]: Indeed, I think it will also conflict with the agent customization test

File: pkg/controllers/provisioningv2/fleetcluster/fleetcluster.go
Comment (manno) [reply pkg/controllers/provisioningv2/fleetcluster/fleetcluster.go:145]: Well yes. This code is replacing the [bootstrap code from Fleet](https://github.com/rancher/fleet/blob/master/pkg/controllers/bootstrap/bootstrap.go) which also ignores user modifications. It's not too bad, since wrangler's apply uses a patch and retains  most modifications.

Fleet's bootstrap would create this group unconditionally for the local cluster, too. Without this special default group every bundle would need an explicit `targets:` section.  We create this group so deployments to the local cluster behave like before. Especially since our fleet-examples repo is full of them.

File: pkg/controllers/provisioningv2/fleetcluster/fleetcluster_test.go
Comment (manno) [reply pkg/controllers/provisioningv2/fleetcluster/fleetcluster_test.go:9]: Is there an attempt to clean up and unify import blocks?

```
% rg '"github.com/rancher/rancher/pkg/apis/management.cattle.io/v3"' | awk '{print $2}' | sort | uniq -c | gsort -rh | head -5                                   
 232 v3
 143 v32
 131 "github.com/rancher/rancher/pkg/apis/management.cattle.io/v3"
  29 apimgmtv3
   4 mgmt
      
% rg '"github.com/rancher/rancher/pkg/apis/provisioning.cattle.io/v1"' | awk '{print $2}' | sort | uniq -c | gsort -rh | head -5
  21 v1
  16 provisioningv1
   8 apisV1
   7 provv1
   7 apiv1

% rg '"github.com/rancher/rancher/pkg/generated/controllers/management.cattle.io/v3"' | awk '{print $2}' | sort | uniq -c | gsort -rh | head -5 
  11 v3
   7 mgmtv3
   6 managementv3
   4 mgmtcontroller
```

Comment (Oats87) [reply pkg/controllers/provisioningv2/fleetcluster/fleetcluster_test.go:9]: @manno that's something we'll scope for Q3, but we could totally add rules to validate aliases are expected values

File: pkg/controllers/provisioningv2/fleetcluster/fleetcluster.go
Comment (Oats87) [inline pkg/controllers/provisioningv2/fleetcluster/fleetcluster.go:13]: nit: there doesn't need to be a newline here

**Review by Oats87 — DISMISSED** [review]

**Review by jakefhyde — APPROVED** [review]
LGTM, but need to resolve merge conflicts

File: pkg/controllers/provisioningv2/fleetcluster/fleetcluster.go
Comment (manno) [reply pkg/controllers/provisioningv2/fleetcluster/fleetcluster.go:22]: I removed the newlines. I normally group imports by "our api/generated", our own pkgs, k8s packages, misc github pkgs, etc.
Helps me keep an overview, but it's hard to enforce in a  project.

**Review by snasovich — APPROVED** [review]
Approving based on previous approval given by @Oats87 and @jakefhyde and trivial changes following those approvals.

EDIT: Also, please do squash merge as there are 4 commits on this PR (or squash and force-push, but this will require re-approval and CI pass again).

**Review by jakefhyde — APPROVED** [review]

**Comment by sbulage** [comment]
I managed to validate few things on latest [Pre-release v2.7.5-rc2](https://github.com/rancher/rancher/releases/tag/v2.7.5-rc2) which has this fix included.

Blocker for me to test it further: Fresh install of Rancher doesn't shows UI.(Known issue, it will be fixed in the next rc build) :crossed_fingers: Hopefully will get sooner.

I had to validate few things on upgraded cluster from `2.7.3 to 2.7.5-rc2` but while testing namespace move, I accidentally moved the `local` cluster to different workspace :disappointed: and hit https://github.com/rancher/fleet/issues/1489.

I am waiting for next RC release to test it.

**Comment by sbulage** [comment]
The [fix](https://github.com/rancher/dashboard/pull/8984) is yet to be added in Rancher latest RC build. Waiting for new RC build to test this issue.

**Comment by sbulage** [comment]
I installed [rancher 2.7.5-rc.4](https://github.com/rancher/rancher/releases/tag/v2.7.5-rc4). 

To verify this issue, need to check below entities:

- [ ] The "local-kubeconfig" secret contains the "internal-server-url" as k8s API server.
- [ ] local-cluster group should not be present as it gets created when local cluster gets bootstrapped by Fleet.
- [ ] Additional `token` field must be present (outside of the `local-kubeconfig` secret)
Selection_028

I kept cluster running almost 18+ hours and I don't see changes happened on the local cluster by fleet-controller specifically.

**Pull request merged**
